cilium/cmd/debuginfo.go

```diff
@@ -77,14 +96,82 @@ var sections = map[string]addSection{
-	debuginfoCmd.Flags().StringVarP(&html, "html-file", "", "", "Convert default output to HTML file")
+	debuginfoCmd.Flags().BoolVarP(&outputToFile, "file", "f", false, "Redirect output to file(s)")
 	debuginfoCmd.Flags().BoolVarP(&filePerCommand, "file-per-command", "", false, "Generate a single file per command")
+	debuginfoCmd.Flags().StringSliceVar(&outputOpts, "output", []string{}, "markdown| html| json| jsonpath='{}'")
+	debuginfoCmd.Flags().StringVar(&outputDir, "output-directory", "", "directory for files (if specified will use directory in which this command was ran)")
+}
+
+func validateInput() {
+	if outputDir != "" && !outputToFile {
+		fmt.Fprintf(os.Stderr, "invalid option combination; specified output-directory %q, but did not specify for output to be redirected to file; exiting\n", outputDir)
+		os.Exit(1)
+	}
+	validateOutputOpts()
+}
+
+func validateOutputOpts() {
+
+	for _, outputOpt := range outputOpts {
+		switch strings.ToUpper(outputOpt) {
+		case markdownOutput:
+		case htmlOutput:
+			if !outputToFile {
+				fmt.Fprintf(os.Stderr, "if HTML is specified as the output format, it is required that you provide the `--file` argument as well\n")
+				os.Exit(1)
+			}
+		case jsonOutputDebuginfo:
+			if filePerCommand {
+				fmt.Fprintf(os.Stderr, "%s does not support dumping a file per command; exiting\n", jsonOutputDebuginfo)
+				os.Exit(1)
+			}
+		case jsonpathOutput:
+			if filePerCommand {
+				fmt.Fprintf(os.Stderr, "%s does not support dumping a file per command; exiting\n", jsonpathOutput)
+				os.Exit(1)
+			}
+		default:
+			// Check to see if arg contains jsonpath filtering as well.
+			if re.MatchString(outputOpt) {
+				return
+			}
+			fmt.Fprintf(os.Stderr, "%s is not a valid output format; exiting\n", outputOpt)
+			os.Exit(1)
+		}
+	}
+}
+
+func formatFileName(outputDir string, cmdTime time.Time, outtype outputType) string {
+	var fileName string
+	var sep string
+	if outputDir != "" {
+		sep = outputDir + "/"
+	}
+	timeStr := cmdTime.Format("20060102-150405.999-0700-MST")
+	switch outtype {
+	case MARKDOWN:
+		fileName = fmt.Sprintf("%scilium-debuginfo-%s.md", sep, timeStr)
+	case HTML:
+		fileName = fmt.Sprintf("%scilium-debuginfo-%s.html", sep, timeStr)
+	case JSONOUTPUT:
+		fileName = fmt.Sprintf("%scilium-debuginfo-%s.json", sep, timeStr)
+	case JSONPATH:
+		fileName = fmt.Sprintf("%scilium-debuginfo-%s.jsonpath", sep, timeStr)
+	default:
+		fileName = fmt.Sprintf("%scilium-debuginfo-%s.md", sep, timeStr)
+	}
+	return fileName
 }
 
 func runDebugInfo(cmd *cobra.Command, args []string) {
-	if os.Getuid() != 0 {
-		fmt.Fprint(os.Stderr, "Warning, some of the BPF commands might fail when run as not root\n")
+
+	validateInput()
+
+	// Only warn when not dumping output as JSON so that when the output of the
+	// command is specified to be JSON, the only outputted content is the JSON
+	// model of debuginfo.
+	if os.Getuid() != 0 && !command.OutputJSON() {
```

Comment:
`command.OutputJSON()` relies upon the usage of `command.AddJSONOutput()`, which I don't see anywhere in this PR. I don't think that this works the way that you expect it to?

Reply:
Good catch.